tracing/trace_state.cc

```diff
@@ -55,15 +55,17 @@ namespace tracing {
 logging::logger trace_state_logger("trace_state");
 
 struct trace_state::params_values {
+    template<typename T, size_t N> using small_vector = ::utils::small_vector<T, N>;
+
     std::optional<std::unordered_set<gms::inet_address>> batchlog_endpoints;
     std::optional<api::timestamp_type> user_timestamp;
-    std::vector<sstring> queries;
+    small_vector<sstring, 1> queries;
     std::optional<db::consistency_level> cl;
     std::optional<db::consistency_level> serial_cl;
     std::optional<int32_t> page_size;
-    std::vector<prepared_checked_weak_ptr> prepared_statements;
-    std::vector<std::optional<std::vector<sstring_view>>> query_option_names;
-    std::vector<std::vector<cql3::raw_value_view>> query_option_values;
+    small_vector<prepared_checked_weak_ptr, 1> prepared_statements;
+    small_vector<std::optional<std::vector<sstring_view>>, 2> query_option_names;
+    small_vector<std::vector<cql3::raw_value_view>, 2> query_option_values;
```

Comment:
The outer vector corresponds to the prepared statements in a batch, and the inner vectors to the query parameters of each statement in a batch. So the outer size should be 1. You should also make the inner vectors small_vectors (but I think there are more parameters in a typical query, let's say 4).

Reply:
Good point. When I was looking at inner vectors I left them for later because I was unsure which part of the code to adapt. The question I have in mind: those _names_ and _values_ they all come from _query_options.hh:77-79_. do we want to change `query_options` struct to have _small_vectors<>_ inside in place of:

```
    const std::optional<std::vector<sstring_view>> _names;
    std::vector<cql3::raw_value> _values;
    std::vector<cql3::raw_value_view> _value_views;
```

or I just rather implement manual copying in the `trace_state` from `optional<vector> -> optional<small_vector>` in

```
void trace_state::add_prepared_query_options(const cql3::query_options& prepared_options_ptr) {
 ...
    for (size_t i = 0; i < _params_ptr->prepared_statements.size(); ++i) {
        _params_ptr->query_option_names.emplace_back(prepared_options_ptr.for_statement(i).get_names());
        _params_ptr->query_option_values.emplace_back(prepared_options_ptr.for_statement(i).get_values());
    }
}
```?

Reply:
query_options.hh will independently benefit from the change to small_vector. We have plenty of areas in the code that would (and with 320f6bf2207259a7ed9a4d4164c70c353054c86f we can even measure the reduced allocations).

Reply:
@avikivity I changed the params struct to have both internal and external vecs as you suggested:

```
template<typename T, size_t N> using small_vector = ::utils::small_vector<T, N>;

using query_names = small_vector<sstring_view, 4>;
using query_values = small_vector<cql3::raw_value_view, 4>;

struct trace_state::params_values {...
    small_vector<prepared_checked_weak_ptr, 1> prepared_statements;
    small_vector<std::optional<query_names>, 1> query_option_names;
    small_vector<query_values, 1> query_option_values;
}
```

However, I didn't rewrite the QueryOptions struct in this patch series. So I had to introduce local helpers for copying:

```
template<typename T, size_t N>
small_vector<T, N> from_vec(const std::vector<T> &v) {
    return small_vector<T, N>(v.begin(), v.end());
}

template<typename T, size_t N>
std::optional<small_vector<T, N>> from_vec(const std::optional<std::vector<T>> &v) {
    if (!v.has_value() || v->empty()) {
        return {};
    }
    return small_vector<T, N>(v->begin(), v->end());
}
```

Are you willing to have a patch for QueryOptions with small_vector in this series or separately?

Reply:
The patch to query_options will be welcome. The new query_names / query_values types should be used throughout, to avoid copying when moving from std::vector to small_vector.

You can prove the reduction in allocations by running perf_simple_query --smp 1 (but see #8550).

Reply:
(better to have it separately)